cynet/client.go

```diff
@@ -0,0 +1,582 @@
+package usp_cynet
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"hash"
+	"hash/fnv"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/refractionPOINT/go-uspclient"
+	"github.com/refractionPOINT/go-uspclient/protocol"
+	"github.com/refractionPOINT/usp-adapters/utils"
+)
+
+const (
+	queryInterval        = 60
+	pageSize             = 100
+	authEndpoint         = "/api/account/token"
+	alertsEndpoint       = "/api/alerts/bulk"
+	maxConsecutiveErrors = 5
+)
+
+type CynetConfig struct {
+	ClientOptions uspclient.ClientOptions `json:"client_options" yaml:"client_options"`
+	// (AccessKey + SecretKey) OR (User + Password)
+	AccessKey string `json:"access_key,omitempty" yaml:"access_key,omitempty"`
+	SecretKey string `json:"secret_key,omitempty" yaml:"secret_key,omitempty"`
+	User      string `json:"user,omitempty" yaml:"user,omitempty"`
+	Password  string `json:"password,omitempty" yaml:"password,omitempty"`
+	SiteID    string `json:"site_id" yaml:"site_id"`
+	URL       string `json:"url,omitempty" yaml:"url,omitempty"` // https://DOMAIN.api.cynet.com
+}
+
+type CynetAdapter struct {
+	conf        CynetConfig
+	uspClient   *uspclient.Client
+	httpClient  *http.Client
+	chStopped   chan struct{}
+	chFetchLoop chan struct{}
+	closeOnce   sync.Once
+	fetchOnce   sync.Once
+	ctx         context.Context
+	cancel      context.CancelFunc
```

Comment:
`accessToken` is written in `authenticate()` and read in `makeRequest()` without synchronization. Consider adding a mutex:
```suggestion
	accessToken   string
	accessTokenMu sync.RWMutex
```
Then use `accessTokenMu.Lock()` in `authenticate()` and `accessTokenMu.RLock()` in `makeRequest()`.